### DISCUSSION TITLE
How to set initial values of variables solved for in study step

### DISCUSSION THREAD

**Discussion by gahorwitz** [MPh-py/MPh#170] [General]
Hello, and thank you for this amazing interface!
I am trying to solve a cyclic time-dependent problem, where the last solution of the 1st step of a study is used as initial values in the second step, and then the last solution for the second step goes back as initial parameters of the first one, and so on.
My problem is that the first time the code is run, the initial values for step 1 must be physics-controlled (instead of the solution of step 2, which doesn't exist yet). My workaround is running the model the first time, then manually changing the initial values to "solution of step 2", and then cycling using mph, but would be great if I could change the initial values from the Python code itself.
 Thank you!

**Comment by john-hen** [comment]
Hi. This is an interesting use case, and certainly one where a scripting interface shines – be it Java, Python, or even Matlab. But since you didn't get a *specific* answer here yet (and I doubt you'll ever get one, given how dormant this "discussion forum" is), rest assured that it's definitely possible to automate your manual work-around. You just need to figure out the API calls, and translate that from Java to Python. Which, for the *general* case, is answered in the pinned discussion, #86 , and by extension also in the documentation.

**Reply by gahorwitz** [reply]
Thank you! I will give it a try that way

**Comment by gahorwitz** [comment]
Thank you! This worked:

```
model.java.study("std1").feature("time").set("useinitsol", True)
model.java.study("std1").feature("time").set("initmethod", "sol")
model.java.study("std1").feature("time").set("initstudy", "std1")
model.java.study("std1").feature("time").set("initsol", "sol1")
model.java.study("std1").feature("time").set("solnum", "last")
```

which changes each line of the initial values, or to start with physics controlled values:

`model.java.study("std1").feature("time").set("useinitsol", False)`

I tried to make it more pythonic but couldn't make it work. I tried variations of (in the last example line)

`model.property("Study//1/time", "useinitsol", False)`

and I get this error: TypeError: eval() arg 1 must be a string, bytes or code object.

and I don't really understand what is the problem.

Thanks again!


``

**Reply by john-hen** [reply]
Thanks for posting the answer!

Yes, that error message is a bit weird. Never noticed that before, maybe I'll improve that at some point. It happens because not only is the node reference wrong (i.e. points to a node that doesn't exist, not by that name anyway), but specifically the "top-level group" is missing. Which is "studies" in this case. Otherwise the error message would be somewhat more helpful.

It should be something like `model.property("studies/Study 1/time dependent", "useinitsol", True)`. Not sure what the study step is called in your model that has the "time" tag in your code. Whatever its name is in the model tree. (Which you see not only in the Comsol GUI, but can also display at the Python prompt with `mph.tree(model)`.)

I'd write it like so, using direct node references instead of `model` methods:
```python
study = model/'studies'/'Study 1'/
step = study/'time dependent'
step.property('useinitsol', True)
step.property('initmethod', 'sol')
step.property('initstudy', study)  # "study" node reference gets replaced by its tag "std1".
step.property('initsol", 'sol1')   # Same should work for solution nodes, just need to know their names.
step.property('solnum', 'last')
```
 Though obviously, as long as it works, that's what matters most.

**Reply by gahorwitz** [reply]
Thank you so much! this works perfectly!
